Add/remove radio encryption keys to the AI Core

Jelkan opened the pull request:
## About The Pull Request

Makes it so players can add/remove radio encryption keys to the AI Core without being able to take out the special AI key

## Why It's Good For The Game

This PR is mainly oriented to ghost roles and servers that use custom/other radio encryption keys.

Players that become off station AI's will no longer need to ahelp to get a encryption key added to their core.
Instead the ones who created the core can now do it, unloading a small burden from the staff team.

This will make it easier for AI to assist with their assigned station incase an admin with sufficient permission isn't available to add the key for them.

Also adds the possibility for others to steal the inserted key, this would still require the AI to either open its core or have it hacked just as previously.

Taking out the special AI radio encryption key is not possible.

## Proof of Testing
<details>
<summary>Picture</summary>
![tgaikey](https://github.com/tgstation/tgstation/assets/20728922/0d2de83e-5fac-4173-9c8a-99f7251640f1)
</details>

## Changelog

:cl:
add: Added possibility to insert/remove radio encryption keys into the AI core
/:cl:

Jelkan committed:
add encryption keys to AI Core

MrMelbert commented:
![image](https://github.com/tgstation/tgstation/assets/51863163/531b85f7-2b6a-4522-a94b-9e6090182835)

MrMelbert commented on code/modules/mob/living/silicon/ai/ai_defense.dm:
```diff
@@ -7,6 +7,14 @@
+		if(radio)
+			radio.attackby(W,user)
+		else
+			to_chat(user, span_warning("Unable to locate a radio!"))
```
Balloon alert

MrMelbert commented on code/modules/mob/living/silicon/robot/robot_defense.dm:
```diff
@@ -105,7 +105,7 @@ GLOBAL_LIST_INIT(blacklisted_borg_hats, typecacheof(list( //Hats that don't real
 
 	if(istype(W, /obj/item/encryptionkey) && opened)
-		if(radio)//sanityyyyyy
+		if(radio && !shell)//removed ability to add key to AI shell, this is done through the AI core.
```
You should add a unique feedback message for shells, rather than `Unable to locate a radio!`. Like `No radio slot!`. Also balloon alert it

Jelkan replied:
adjusted to your request

MrMelbert commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -358,15 +358,15 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 /obj/item/radio/headset/screwdriver_act(mob/living/user, obj/item/tool)
 	user.set_machine(src)
-	if(keyslot || keyslot2)
+	if(keyslot || keyslot2 && !istype(keyslot2, /obj/item/encryptionkey/ai)) //Check to make sure you cant remove the AI key from core 
```
I don't really like this hardcode istype, perhaps there should be a trait or variable that denotes an encryption key as unremovable ? If i'm understanding the purpose of the code correctly

Jelkan replied:
Purpose is really only to make sure that players cant remove specifically the AI key, since this one has all station channels and ai private.
If i have understood it correctly, its a key that IC doesnt exist and should not be able to end up in a players hand.

Not sure if there is a better way to do it, but I do welcome feedback and ideas.

MrMelbert replied:
I understand your purpose but I would rather we had a variable or something to denote `this key should never be removed from headsets as players should not have their hands on it` 
...Rather than checking for this type directly

Jelkan replied:
That actually makes a lot of sense and shouldn't be too difficult to work in to set if a key can or cant be removed.
I will see what magic I can work up tomorrow.

Jelkan committed:
Update ai_defense.dm

Jelkan committed:
Update ai_defense.dm

Changed from to_chat to balloon alert

Jelkan committed:
Update robot_defense.dm

added unique message incase its a shell they are trying to remove a key from

Jelkan commented on code/modules/mob/living/silicon/ai/ai_defense.dm:
```diff
@@ -160,3 +168,15 @@
+		if(radio)
+			radio.screwdriver_act(user, tool)
+		else
+			to_chat(user, span_warning("Unable to locate a radio!"))
```
also changed

Jelkan commented:
> ![image](https://user-images.githubusercontent.com/51863163/244475297-531b85f7-2b6a-4522-a94b-9e6090182835.png)

i dont know why, but github doesnt seem to like me drag and dropping images and i am not sure what is wrong or what i am doing wrong or how to fix it, still rather new on the whole github thing.

MrMelbert commented on code/modules/mob/living/silicon/ai/ai_defense.dm:
```diff
@@ -7,6 +7,14 @@
+		if(radio)
+			radio.attackby(W,user)
+		else
+			balloon_alert(user,"Unable to locate a radio!")
```
```suggestion
			balloon_alert(user, "no internal radio!")
```

MrMelbert commented on code/modules/mob/living/silicon/ai/ai_defense.dm:
```diff
@@ -160,3 +168,15 @@
+		if(radio)
+			radio.screwdriver_act(user, tool)
+		else
+			balloon_alert(user, "Unable to locate a radio!")
```
```suggestion
			balloon_alert(user, "no internal radio!")
```

MrMelbert commented on code/modules/mob/living/silicon/robot/robot_defense.dm:
```diff
@@ -105,8 +105,10 @@ GLOBAL_LIST_INIT(blacklisted_borg_hats, typecacheof(list( //Hats that don't real
+		if(radio && !shell)//removed ability to add key to AI shell, this is done through the AI core.
 			radio.attackby(W,user)//GTFO, you have your own procs
+		else if(shell)
+			balloon_alert(user, "No radio slot!")
```
```suggestion
			balloon_alert(user, "no radio key slot!")
```

MrMelbert commented on code/modules/mob/living/silicon/robot/robot_defense.dm:
```diff
@@ -105,8 +105,10 @@ GLOBAL_LIST_INIT(blacklisted_borg_hats, typecacheof(list( //Hats that don't real
+		else if(shell)
+			balloon_alert(user, "No radio slot!")
 		else
 			to_chat(user, span_warning("Unable to locate a radio!"))
```
This one too for consistency

MrMelbert replied:
Don't forget this

Jelkan committed:
Update code/modules/mob/living/silicon/robot/robot_defense.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/modules/mob/living/silicon/ai/ai_defense.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/modules/mob/living/silicon/ai/ai_defense.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Merge branch 'tgstation:master' into master

Jelkan committed:
Add radio keys to AI core

Players can now add radio keys to the AI core and added var to set radiokeys as unremovable

MrMelbert commented on code/game/objects/items/devices/radio/encryptionkey.dm:
```diff
@@ -10,6 +10,8 @@
 	var/syndie = FALSE
 	/// If true, the radio can say/hear on the special CentCom channel.
 	var/independent = FALSE
+	/// Determines if the key can be removed from a radio/headset
+	var/canRemove = TRUE
```
`snake_case`

Jelkan replied:
hmm, not sure if that means you liked it, tagged someone or want something changed

MrMelbert commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -358,23 +358,23 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 /obj/item/radio/headset/screwdriver_act(mob/living/user, obj/item/tool)
 	user.set_machine(src)
-	if(keyslot || keyslot2)
+	if(keyslot && keyslot.canRemove || keyslot2 && keyslot2.canRemove)
```
```suggestion
	if(keyslot?.canRemove || keyslot2?.canRemove)
```

MrMelbert commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -358,23 +358,23 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 			secure_radio_connections[ch_name] = null
 
-		if(keyslot)
+		if(keyslot && keyslot.canRemove)
```
```suggestion
		if(keyslot?.canRemove)
```

MrMelbert commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -358,23 +358,23 @@ GLOBAL_LIST_INIT(channel_tokens, list(
 			user.put_in_hands(keyslot)
 			keyslot = null
-		if(keyslot2)
+		if( keyslot2 && keyslot2.canRemove)
```
```suggestion
		if(keyslot2?.canRemove)
```

MrMelbert commented on code/game/objects/items/devices/radio/headset.dm:
```diff
@@ -358,23 +358,23 @@ GLOBAL_LIST_INIT(channel_tokens, list(
-
 	else
-		to_chat(user, span_warning("This headset doesn't have any unique encryption keys! How useless..."))
+		to_chat(user, span_warning("This headset doesn't have any encryption keys!"))
```
Should include a detail if there are keys, just un-removable ones (Because I think they still show up on examine? So it might be misleading) -
```suggestion
		to_chat(user, span_warning("This headset doesn't have any[keyslot || keyslot2 ? " removable" : ""] encryption keys!"))
```

MrMelbert commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -545,23 +545,22 @@
-	for(var/ch_name in channels)
-		SSradio.remove_object(src, GLOB.radiochannels[ch_name])
-		secure_radio_connections[ch_name] = null
+	if(keyslot && keyslot.canRemove)
```
```suggestion
	if(keyslot?.canRemove)
```

MrMelbert commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -545,23 +545,22 @@
+		to_chat(user, span_notice("You pop out the encryption key in the radio."))
+		return
+	else
+		to_chat(user, span_warning("This radio doesn't have any encryption keys!"))
```
Same as above with "any removable keys"

Jelkan replied:
Alright, i added that here as well, shorten it to only keyslot since you cant insert anything into keyslot2 on the radio

MrMelbert commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -545,23 +545,22 @@
-	to_chat(user, span_notice("You pop out the encryption key in the radio."))
-	return ..()
+		recalculateChannels()
+		to_chat(user, span_notice("You pop out the encryption key in the radio."))
+		return
+	else
+		to_chat(user, span_warning("This radio doesn't have any encryption keys!"))
+		return
```
These two returns are pointless. Also, this used to call parent, now it doesn't. Was that intentional?

Jelkan replied:
This was intentional yes, when the parent was called, you would open up the cyborgs radio as well as remove the key, but you cant access the radios wires with neither wirecutter or multitool, but if you are the cyborg, opening your radio opens up the wire window.
I dont know what the original coder of the cyborg radio had intended with it or if they missed adding something, so I decided to no longer call the parent.

the returns i will remove

Jelkan committed:
Update code/game/objects/items/devices/radio/headset.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/game/objects/items/devices/radio/headset.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/game/objects/items/devices/radio/headset.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/game/objects/items/devices/radio/headset.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update code/game/objects/items/devices/radio/radio.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jelkan committed:
Update radio.dm

Jelkan committed:
Update radio.dm

Jelkan committed:
Update robot_defense.dm

Jelkan committed:
Update radio.dm

Fix of missing "

Jelkan committed:
Update radio.dm

Jelkan committed:
Update radio.dm

removed variable message since it only has one slot

Jelkan committed:
Update robot_defense.dm

fixed typo

Jelkan commented:
Okay, finally had time to fix my silly little mistakes I had made :)
